Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs

```diff
@@ -86,39 +86,49 @@ public virtual void ProcessBlock(Block block)
 
             // If the new block previous hash is the same as the 
             // wallet hash then just pass the block to the manager. 
+            var newTip = this.chain.GetBlock(block.GetHash());
+
+            // new block is not in the main chain nothing more to do right now. 
+            if (newTip == null)
+                return; // reorg
+
             if (block.Header.HashPrevBlock != this.walletTip.HashBlock)
             {
-                // set a temporary wallet tip in case we are not on the best chain anymore.
-                var walletTip = this.WalletTip;
-
                 // If previous block does not match there might have 
                 // been a reorg, check if the wallet is still on the main chain.
                 ChainedBlock inBestChain = this.chain.GetBlock(this.walletTip.HashBlock);
                 if (inBestChain == null)
                 {
                     // The current wallet hash was not found on the main chain.
-                    // A reorg happenend so bring the wallet back top the last known fork.
+                    // A reorg happened so bring the wallet back top the last known fork.
 
-                    var fork = walletTip;
+                    var fork = this.walletTip;
 
                     // We walk back the chained block object to find the fork.
                     while (this.chain.GetBlock(fork.HashBlock) == null)
                         fork = fork.Previous;
```

Comment:
In general, I find it very difficult to get my brain grasping this loop. Why? Consider how much `this.chain` can change in this method and in this loop. At any point in time in this loop, `this.chain` can be different chain. So we might start with chain 1:
```
A-B-C-D-E-F-G
```

and we just jumped onto different chain 2:

```
A-B-C-D-E2-F2-G2-H2
```
So, the wallet tip when we entered the method was G, so we are not executing this loop to find the fork - we are hoping to find D. So we go from G to F, but then the chain switches to chain 3:

```
A-B3-C3-D3-E3-F3-G3-H3-I3
```

we know nothing, so we continue down the chain to E, then to D, then to C and now the chain switches again to 

```
A-B-C-D-E2-F2-G2-H2-I2-J2
```

so we go to B instead of D. Not sure if that matters for further execution of the method and the logic behind it, I wrote it just to demonstrate how difficult is to think about this loop. So at the end we could arrive anywhere really and once we finish this loop, we don't have any guarantees that `fork` is actually in our chain. 

So I don't say there is a bug in this code, I just say that I'm very close to be unable to think about it clearly as there are too many ways how this can play out. This does not seem like the correct design, it is just too complex to think about (for me).

Reply:
You have valid points, and the solution to this is making the wallet stand alone, what does that mean? the wallet not depending on consensus (like with store, consensus will feed blocks to the wallet but if the blocks are out of band wallet will handle that itself without stopping consensus by either asking from store or downloading form peers).

This will be the best approach.

> So at the end we could arrive anywhere really and once we finish this loop, we don't have any guarantees that fork is actually in our chain.
This is true for any given moment right? as the chain could in theory reorg at any point, best we can do is finish (fully/partially) our current reorg operation and next round try to fix it.

I think the most important is that the wallet can recover from any scenario, if we end up on the wrong chain after all this work, the next block will fix that and the wallet will recover without losing fund.

Reply:
Yes, at no given point you have any guarantees about the value in local variable representing something on the best chain.

I will try to present a concurrent chain replacement idea "soon".

But yes, you just need to be able to recover. This includes not crashing the node or loop you are in and possibly not executing too much complex logic with invalid values, which can be tricky.

Reply:
I will remind you that this will be solved much better if the wallet was a stand alone feature.
It would run in the background and only make one check once in a while that it's on the best chain (and if its not reorg and rebuild the wallet without depending on he consensus feeding blocks, either by downloading blocks or read from store)

Reply:
Yes, that is excellent point. As mentioned, maybe it is not worth fixing it now because of this upcoming change that will dramatically change the whole thing.